File: lib/AdventOfCode/Solution/Day04.ex
defmodule Solutions.Day04 do
  alias Helpers.Readers, as: Reader

  def checkRedundant(input) do
    [min1, max1, min2, max2] = input
    cond do
      (min1 <= min2 && max1 >= max2) || (min2 <= min1 && max2 >= max1) -> 1
      true -> 0
    end
  end

  def checkRedundantT2(input) do
    [min1, max1, min2, max2] = input
    cond do
      (max1 >= min2 && max1 <= max2) || (max2 >= min1 && max2 <= max1) -> 1
      true -> 0
    end
  end

  def run(args) do
      data = Reader.readFile(args)
      |> Enum.map(fn x -> String.replace(x, "-", ",") end)
      |> Enum.map(fn x -> String.split(x, ",") end)
      |> Enum.map(fn x -> Enum.map(x, &String.to_integer/1) end)

      Enum.map(data, &checkRedundant/1)
      |> Enum.sum
      |> IO.inspect(label: "Task 01")

      Enum.map(data, &checkRedundantT2/1)
      |> Enum.sum
      |> IO.inspect(label: "Task 02")
  end
end

File: src/main/scala/Solutions/Day04.scala
package AoC2022
package Solutions

import Helpers.Readers

object Day04 {
  def main(args: Array[String]): Unit = {
    val data = Readers.readFile("Data/Day04.txt")
    val parsed = data.map(_.replace(",", "-").split("-").map(_.toInt).toList).map(new Limits(_))
    println("Task 01: " + parsed.map(_.overlap()).count(_ == true))
    println("Task 02: " + parsed.map(_.overlapAny()).count(_ == true))
  }

  case class Limits(min1: Int, max1: Int, min2: Int, max2: Int){
    def this(entry: List[Int]) = this(entry.head,entry(1), entry(2), entry(3))
    def overlap() : Boolean = {
      (min1 >= min2 && max1 <= max2) || (min2 >= min1 && max2 <= max1)
    }
    def overlapAny() : Boolean = {
      (max1 >= min2 && max1 <= max2) || (max2 >= min1 && max2 <= max1)
    }
  }
}
